### PR TITLE
Update MX beamlines to use new device factory - part 1

### PR DESCRIPTION
**Pull request by rtuck99**
Fixes 
* #846

for I03, other beamlines to follow in separate PR on top of this one.

Also fixes the following issues:
* `diamond_filter` was not being instantiated because python generic device types were not recognised as valid device factories.
* ~`device_factory` did not support usage with sync v1 ophyd devices~ These changes now in follow on PR #984 

See also mx-bluesky PR
* DiamondLightSource/mx-bluesky#735

### Instructions to reviewer on how to test:
1. Dodal connect i03, i04 should still work

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://diamondlightsource.github.io/dodal/main/reference/device-standards.html)
- [ ] If changing the API for a pre-existing device, ensure that any beamlines using this device have updated their Bluesky plans accordingly
- [ ] Have the connection tests for the relevant beamline(s) been run via `dodal connect ${BEAMLINE}`

### CODE REVIEW

**Review by olliesilvester — CHANGES_REQUESTED** [review]
Looks good, but I feel like we should put the part which fixes `device_factory` for v1 devices into a separate PR. 

Also, for this issue we might as well convert all of the devices in i04, i24 and i23 to use the new instantiation

File: src/dodal/beamlines/i03.py [hunk 1]
```diff
@@ -57,494 +57,367 @@
...
+        f"{PREFIX.beamline_prefix}-MO-DCM-01:",
         "dcm",
-        "-MO-DCM-01:",
-        wait_for_connection,
-        fake_with_ophyd_sim,
     )
 
 
+@device_factory()
 @skip_device(lambda: BL == "s03")
```
Comment (DiamondJoseph) [inline src/dodal/beamlines/i03.py:97]: Depending on how mx-bluesky/hyperion is handling this:

```suggestion
@device_factory(skip = BL == "s03")
```

Comment (DiamondJoseph) [reply src/dodal/beamlines/i03.py:97]: or `skip = (BL == "s03")` if it's any clearer what is meant

File: src/dodal/beamlines/i03.py [hunk 2]
```diff
@@ -57,494 +57,367 @@
...
     If this is called when already instantiated in i03, it will return the existing object.
     """
     return device_instantiation(
         Flux,
         "flux",
         "-MO-FLUX-01:",
-        wait_for_connection,
-        fake_with_ophyd_sim,
+        wait=False,
+        fake=mock,
```
Comment (DiamondJoseph) [inline src/dodal/beamlines/i03.py:321]: Seems to be mixing the two device factory types?

File: src/dodal/common/beamlines/beamline_utils.py [hunk 3]
```diff
@@ -141,8 +140,8 @@ def device_factory(
         SkipType,
         "mark the factory to be (conditionally) skipped when beamline is imported by external program",
     ] = False,
-) -> Callable[[Callable[[], D]], DeviceInitializationController[D]]:
-    def decorator(factory: Callable[[], D]) -> DeviceInitializationController[D]:
+) -> Callable[[Callable[[], T]], DeviceInitializationController[T]]:
+    def decorator(factory: Callable[[], T]) -> DeviceInitializationController[T]:
```
Comment (DiamondJoseph) [inline src/dodal/common/beamlines/beamline_utils.py:144]: This has to be D, Ophyd-Async only device, the connection logic does not work for Ophyd devices.

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]

File: src/dodal/common/beamlines/beamline_utils.py (hunk 3 above)
Comment (rtuck99) [reply src/dodal/common/beamlines/beamline_utils.py:144]: I have made modifications so that the device factory will now work with ophyd v1 devices.

Comment (DiamondJoseph) [reply src/dodal/common/beamlines/beamline_utils.py:144]: So you have, if you can't tell that I stopped looking when I reached this point ;)

I thought MX only had the Eiger left on ophyd v1?

Comment (rtuck99) [reply src/dodal/common/beamlines/beamline_utils.py:144]: Plus a few other minor devices, s4_slits, etc. There is currently a GH ticket to move those across but the Eiger is the main thing,  it will probably be a little while yet before that gets moved over to ophyd_async. I thought better to make a start on this.

Once they are moved over, we will be able to get rid of a whole bunch of code including the `device_instantiation` stuff.

File: src/dodal/beamlines/i03.py (hunk 1 above)
Comment (rtuck99) [reply src/dodal/beamlines/i03.py:97]: At this point, I think we are likely going to remove all this `s03` stuff soon as I think we have pretty much given up on the simulator.

Comment (rtuck99) [reply src/dodal/beamlines/i03.py:97]: There is a separate issue to remove the s03 system tests and I have updated that to remove s03 completely

* DiamondLightSource/mx-bluesky#631

**Comment by rtuck99** [comment]
> Looks good, but I feel like we should put the part which fixes `device_factory` for v1 devices into a separate PR.

These changes now in 
* #984 

> Also, for this issue we might as well convert all of the devices in i04, i24 and i23 to use the new instantiation

These changes to follow in subsequent PRs

File: src/dodal/beamlines/i04.py [hunk 4]
```diff
@@ -427,17 +429,13 @@ def oav_to_redis_forwarder(
     )
 
 
-def diamond_filter(
-    wait_for_connection: bool = True, fake_with_ophyd_sim: bool = False
-) -> DiamondFilter[I04Filters]:
+@device_factory()
+def diamond_filter() -> DiamondFilter[I04Filters]:
```
Comment (olliesilvester) [inline src/dodal/beamlines/i04.py:433]: Doesn't really matter, but did you mean to take out this change in favour for your i04 PR?

**Review by olliesilvester — CHANGES_REQUESTED** [review]
The i03 changes look good, but I think there are some lingering changes which you have also made in some other PRs, would you mind taking them out of this one so that we don't end up merging the wrong features from this PR?

File: src/dodal/utils.py [hunk 5]
```diff
@@ -410,7 +412,27 @@ def is_any_device_factory(func: Callable) -> TypeGuard[AnyDeviceFactory]:
 
 
 def is_v2_device_type(obj: type[Any]) -> bool:
```
Comment (DominicOram) [inline src/dodal/utils.py:414]: Should: Would be nice to have a test that explicitly checks this works for generics.
Could: Also testing the edge cases of not passing a class etc.

**Review by DominicOram — CHANGES_REQUESTED** [review]
Thanks, just one comment

File: src/dodal/beamlines/i04.py (hunk 4 above)
Comment (rtuck99) [reply src/dodal/beamlines/i04.py:433]: If these lines aren't updated, then `test_device_creation` fails without further modifications because `device_instantiation` also doesn't work with generics, so then I need to fix `device_instantiation` as well.

File: tests/test_utils.py [hunk 6]
```diff
@@ -19,30 +21,59 @@
     get_beamline_based_on_environment_variable,
     get_hostname,
     get_run_number,
+    is_v2_device_type,
     make_all_devices,
     make_device,
 )
 
+# Duplicated here because of top-level import issues
```
Comment (olliesilvester) [inline tests/test_utils.py:29]: Is there an issue for this we can link to in the comment?

**Review by olliesilvester — APPROVED** [review]
I'm happy previous comments have been addressed, although I don't fully understand what's going in `is_v2_device_type`
